### PR TITLE
Add new implementation for RCC

### PR DESCRIPTION
**Pull request by no111u3**
Create new version of RCC setup, with direct setup clocks.
Also update clock calculation and add usb pll configuration.

### CODE REVIEW

File: src/rcc.rs [hunk 1]
```diff
@@ -164,231 +166,117 @@ impl HSEClock {
 
 const HSI: u32 = 16_000_000; // Hz
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum PLLP {
+    Pllp2 = 0b00,
+    Pllp4 = 0b01,
+    Pllp6 = 0b10,
+    Pllp8 = 0b11,
```
Comment (Disasm) [inline src/rcc.rs:174]: It's better to call them `Div2`, `Div4`, ...

File: src/rcc.rs [hunk 2]
```diff
@@ -164,231 +166,117 @@ impl HSEClock {
...
+                    PLLP::Pllp2 => 2,
+                    PLLP::Pllp4 => 4,
+                    PLLP::Pllp6 => 6,
+                    PLLP::Pllp8 => 8,
+                };
 
-            use_pll = true;
+            if self.use_pll48clk {
+                pll48clk_valid =
+                    base_clk / self.pllm as u32 * self.plln as u32 / self.pllq as u32 == 48_000_000;
```
Comment (Disasm) [inline src/rcc.rs:269]: According to the f4 hal code, USB specification allows +-0.25%

File: src/rcc.rs [hunk 3]
```diff
@@ -164,231 +166,117 @@ impl HSEClock {
...
-
-        // Switch to fail-safe clock settings.
-        // This is useful when booting from a bootloader that alters clock tree configuration.
-        // Turn on HSI
-        rcc.cr.modify(|_, w| w.hsion().set_bit());
-        while rcc.cr.read().hsirdy().bit_is_clear() {}
-        // Switch to HSI
-        rcc.cfgr.modify(|_, w| w.sw().hsi());
+    /// Output clock calculation
+    fn calculation_clocks(&self) -> (Clocks, InternalRCCConfig) {
```
Comment (Disasm) [inline src/rcc.rs:246]:
```suggestion
    fn calculate_clocks(&self) -> (Clocks, InternalRCCConfig) {
```

File: src/rcc.rs [hunk 4]
```diff
@@ -164,231 +166,117 @@ impl HSEClock {
...
-                });
-            }
+        let mut pll48clk_valid = false;
 
-            // Enable PLL
-            rcc.cr.modify(|_, w| w.pllon().on());
-            // // Wait for PLL to stabilise
-            while rcc.cr.read().pllrdy().is_not_ready() {}
+        if self.use_pll {
+            sysclk = base_clk / self.pllm as u32 * self.plln as u32
```
Comment (Disasm) [inline src/rcc.rs:259]:
```suggestion
            sysclk = base_clk * self.plln as u32 / self.pllm as u32
```

File: src/rcc.rs (hunk 1 above)
Comment (no111u3) [reply src/rcc.rs]: Done.

File: src/rcc.rs (hunk 2 above)
Comment (no111u3) [reply src/rcc.rs]: Done.

File: src/rcc.rs [hunk 5]
```diff
@@ -478,28 +450,276 @@ impl CFGR {
...
+
+        Some((m, n, p, q))
+    }
+
+    /// Configure system clock settings by providen user clock frequency
+    /// Set sysclk as providen and setup USB clock if defined.
+    /// It must be between 12.5 Mhz and 216 Mhz.
+    /// If the ethernet peripheral is on, the user should set a
+    /// frequency higher than 25 Mhz
+    pub fn set_sysclk<F>(self, sysclk: F) -> Option<Self>
```
Comment (mvertescher) [inline src/rcc.rs:580]: Maybe something like:

```
    /// Configure the system clock settings. 
    ///
    /// This sets the sysclk frequency and sets up the USB clock if defined.
    /// The provided frequency must be between 12.5 Mhz and 216 Mhz.
    /// If the ethernet peripheral is on, the user should set a frequency higher than 25 Mhz.
```

File: src/rcc.rs [hunk 6]
```diff
@@ -478,28 +450,276 @@ impl CFGR {
...
+            });
+
+            if config.overdrive {
+                pwr.cr1.modify(|_, w| w.oden().set_bit());
+            }
+
+            // Enable PLL
+            rcc.cr.modify(|_, w| w.pllon().on());
+
+            // // Wait for PLL to stabilise
```
Comment (mvertescher) [inline src/rcc.rs:698]: extra `//` here, maybe:  `// Wait for PLL to stabilise`

**Comment by mvertescher** [comment]
We should do some testing on this PR before it lands just in case. If anyone has a chance to run through some sanity configurations that would be helpful. I may have some time later this week to test, unsure though.

File: src/rcc.rs (hunk 5 above)
Comment (no111u3) [reply src/rcc.rs:580]: Best variant, thanks.

File: src/rcc.rs (hunk 6 above)
Comment (no111u3) [reply src/rcc.rs]: Sure, removed.

File: src/rcc.rs [hunk 7]
```diff
@@ -220,175 +269,85 @@ impl CFGR {
...
-            use_pll = true;
+        if self.use_pll48clk {
+            pll48clk_valid = {
+                let pll48clk =
+                    base_clk as u64 * self.plln as u64 / self.pllm as u64 / self.pllq as u64;
+                pll48clk >= 48_000_000 + 120_000 && pll48clk <= 48_000_000 + 120_000
+            };
         }
+        // SYSCLK, must be <= 216 Mhz. By default, HSI/HSE frequency is chosen
+        assert!(sysclk <= 216);
```
Comment (Disasm) [inline src/rcc.rs:343]:
```suggestion
        assert!(sysclk <= 216_000_000);
```

**Comment by Disasm** [comment]
Result of calculations depends on the position of `.set_sysclk()` call in a builder chain, this is not good.

File: src/rcc.rs [hunk 8]
```diff
@@ -618,17 +618,15 @@ impl CFGR {
             if let Some(q) = q {
                 self.pllq(q as u8);
             }
-
-            Some(self)
-        } else {
-            None
         }
```
Comment (Disasm) [inline src/rcc.rs:621]: else panic?

File: src/rcc.rs (hunk 7 above)
Comment (no111u3) [reply src/rcc.rs]: Done.

File: src/rcc.rs (hunk 8 above)
Comment (no111u3) [reply src/rcc.rs:623]: Done.

**Comment by no111u3** [comment]
> Result of calculations depends on the position of `.set_sysclk()` call in a builder chain, this is not good.

Fixed.

**Comment by no111u3** [comment]
> Also this test fails:
> 
> ```rust
>     #[test]
>     fn test_rcc_calc1() {
>         use super::{PLLP, HSEClock, HSEClockMode};
>         use crate::time::U32Ext;
> 
>         let cfgr = CFGR {
>             hse: None,
>             hclk: None,
>             pclk1: None,
>             pclk2: None,
>             use_pll: false,
>             use_pll48clk: false,
>             pllm: 2,
>             plln: 50,
>             pllp: PLLP::Div2,
>             pllq: 2,
>         };
> 
>         let cfgr = cfgr
>             .hse(HSEClock::new(25.mhz(), HSEClockMode::Bypass))
>             .use_pll()
>             .use_pll48clk()
>             .sysclk(216.mhz());
> 
>         let (clocks, config) = cfgr.calculate_clocks();
>         assert_eq!(clocks.sysclk().0, 216_000_000);
>         assert!(clocks.is_pll48clk_valid());
>     }
> ```

Fixed, test added to test group in rcc.

**Comment by mvertescher** [comment]
@Disasm any other thoughts on this PR? If not, I'll merge this and we can follow up with any issues that arise.

File: src/rcc.rs [hunk 9]
```diff
@@ -478,28 +470,272 @@ impl CFGR {
...
+    ///
+    /// The implementation makes the following choice: HSI is always chosen over
+    /// HSE except when HSE is provided. When HSE is provided, HSE is used
+    /// wherever it is possible.
+    pub fn freeze(self) -> Clocks {
+        let flash = unsafe { &(*FLASH::ptr()) };
+        let rcc = unsafe { &(*RCC::ptr()) };
+        let pwr = unsafe { &(*PWR::ptr()) };
+
+        self.pll_configure();
```
Comment (Disasm) [inline src/rcc.rs:664]: `pll_configure` is Self->Self, so this operation is no-op.

Comment (no111u3) [reply src/rcc.rs:664]: Fixed

File: src/rcc.rs [hunk 10]
```diff
@@ -656,7 +654,7 @@ impl CFGR {
     /// The implementation makes the following choice: HSI is always chosen over
     /// HSE except when HSE is provided. When HSE is provided, HSE is used
     /// wherever it is possible.
-    pub fn freeze(self) -> Clocks {
+    pub fn freeze(&mut self) -> Clocks {
```
Comment (Disasm) [inline src/rcc.rs:657]: I think it's a bad idea...

Comment (no111u3) [reply src/rcc.rs:657]: Fixed.

**Review by Disasm — APPROVED** [review]
I tested 4c9af734cdb0b98cd60eecd873a038109a0f7c89 within this PR on my STLinkV3-SET as target board and both rcc and USB serial tests passed. Let's see what CI tells us...

**Comment by Disasm** [comment]
Apparently, tests should be updated to reflect `pll_configure()` signature change.

**Comment by no111u3** [comment]
> Apparently, tests should be updated to reflect `pll_configure()` signature change.

Fixed.

**Review by mvertescher — APPROVED** [review]

**Pull request merged**